cql3: Fix reverse-order collections

dekimir opened the pull request:
When a clustering order is reversed, the corresponding column type is set to reversed_type_impl.  This was not properly handled by collections code, which assumed the type is a regular collection implementation.  Fixed by getting the underlying type when appropriate.

Fixes #7868
Fixes #7875

Tests: unit (dev)

dekimir commented:
@nyh I added my own test, since I couldn't apply your patch to master.  Sorry. :(

psarna commented:
(I took liberty of adding `Fixes #7868` to the description, since that would get translated to the merged commit message - it's easier to backport if the merge message also mentions the issue.)

psarna reviewed (approved):
Looks good to me, but from the discussion I conclude that there's another conflicting patch which adds tests, so I'll wait for @nyh to confirm

nyh commented on cql3/abstract_marker.cc:
```diff
@@ -69,18 +69,19 @@ abstract_marker::raw::raw(int32_t bind_index)
 {
-    if (receiver->type->is_collection()) {
-        if (receiver->type->get_kind() == abstract_type::kind::list) {
+    const auto tp = receiver->type->underlying_type();
```
Wow. So potentially we have such type->get_kind() calls returning the wrong thing for reversed collections all over the code. Would be good if you or someone else could audit for that...
I'm curious - why can't type->get_kind() return the "kind" of the underlying type, to solve this problem for all callers?

nyh replied:
Also, a question: I understand what underlying_type does for reversed-order types. Is it possible that for other things, it has a different meaning, so it wouldn't be correct to use underlying_type?

dekimir replied:
> Is it possible that for other things, it has a different meaning, so it wouldn't be correct to use underlying_type?

That's definitely possible, which is why I had to add the extra is_reversed() check in lists.cc.  I'll add it everywhere, just to be sure.

dekimir replied:
I added two utility methods to abstract_type: self_or_reversed() and as().  But they must be called explicitly.

nyh commented:
Oh, and please check using my v2 tests if your patch also fixes #7875, and if not, consider a patch which fixes both bugs (if it's not the same problem, it's probably almost the same fix, with the underlying_type thing).

dekimir commented on test/cql-pytest/cassandra_tests/validation/entities/collections_test.py:
```diff
@@ -1583,3 +1583,10 @@ def testSelectionOfEmptyCollections(cql, test_keyspace):
             assert_rows(execute(cql, table, "SELECT m['0'..'1']['3'..'5'], s[0..1][3..5] FROM %s WHERE k = 3"), [None, None])
+
+# Regression-test for #7868.
+def test_7868(cql, test_keyspace):
```
Use it how?  Wait for it to land in master?  Or wholy incorporate it here?

nyh replied:
I meant apply it locally just to see my tests pass with your new code. You don't need to include it inside your patch.
But whatever you do, you can't add new tests to the cassandra_tests/ subdirectory - those are supposed to be translated tests, not new tests :-)

dekimir replied:
Good point about translated-vs-new tests.  But I do want to have _a_ test with the code change.  Either your test patch has to get in (so mine will just remove `@skip`) or the two should be united into one.

dekimir replied:
In v2, I didn't add any tests (though I tested with the tests from the mailing list).

dekimir commented:
> Oh, and please check using my v2 tests if your patch also fixes #7875

Already checked; it doesn't.

> and if not, consider a patch which fixes both bugs

I considered it and rejected it, because I want to prioritize crashes over lesser bugs.  Crashes should be fixed promptly, without waiting for extra work elsewhere.

nyh commented:
> > Oh, and please check using my v2 tests if your patch also fixes #7875
> 
> Already checked; it doesn't.
> 
> > and if not, consider a patch which fixes both bugs
> 
> I considered it and rejected it, because I want to prioritize crashes over lesser bugs. Crashes should be fixed promptly, without waiting for extra work elsewhere.

I think this is a mistake... You're already thinking about this code, I already wrote tests for it - why not just fix it? The fix is probably (but I'm just guessing) yet another test for "kind" which should look at underlying_type...

dekimir commented:
> > I considered it and rejected it, because I want to prioritize crashes over lesser bugs. Crashes should be fixed promptly, without waiting for extra work elsewhere.
> 
> I think this is a mistake... You're already thinking about this code, I already wrote tests for it - why not just fix it?

Because it's not this code, it's some other code that I want to debug separately from this.  I have a fix for the crash, and I don't understand why that should wait for some other work in another part of the code.

nyh commented:
> > I think this is a mistake... You're already thinking about this code, I already wrote tests for it - why not just fix it?
> 
> Because it's not this code, it's some other code that I want to debug separately from this. I have a fix for the crash, and I don't understand why that should wait for some other work in another part of the code.

During the hours we have spent on writing these tests and debugging the problem, we discovered a serial bug - that every piece of code that looks if "type->kind == list" will miss reversed collections. There may be a dozen such bugs lurking in the program, but we know of at least *two*. If you don't want to spend your time look for all of them, you should at least look at the two we know about. The second one is `sets::literal::validate_assignable_to` and friends, which also check type without considering a reversed type's underlying_type.

There are two reasons why I insist that you look at both problems together:

1. Since we already did 99% of the work of finding the cause and writing tests, let's finish it!
2. After looking at the two problems, maybe this will influence how you think we should solve this problem. Maybe we should have some function that both pieces of code will use?

I think that making "reversed" a wrapper type (done 5 years ago in commit 1047f36a9a391ad75bd127903928309bbf21c0f7) might have been a mistake, it complicates all other code which needs to understand the real type. I'm not asking you to undo this, but I'm asking you to think a little bit further beyond fixing a specific crash.


It's not "some other work in another part of the code". I am guessing (and if you don't even look, you can't disprove my guess) th

dekimir commented:
@nyh I honestly didn't think it would take so many hours to commit a clean crash fix.  But now that it's happened, I guess I'll spend some more time to fix both bugs in one patchset and delay my other work. :man_shrugging:

nyh commented:
> @nyh I honestly didn't think it would take so many hours to commit a clean crash fix. But now that it's happened, I guess I'll spend some more time to fix both bugs in one patchset and delay my other work. man_shrugging

You feel this way just because you are underestimating the amount of work that *I* did to discover, document, and test this bug. These are not two "unrelated" bugs - they are places in the code with exactly the same bug - checking "type" without considering underlying() as well. Since we already spent so much time to understand this, now is the best time to fix both places, not just one... (and seeing your objection, I'll not insist that you look for more instances of the same bug, though I suspect there are...).

dekimir commented:
I pushed v2, PTAL.  I used the tests from the mailing list, though I didn't include them in this PR.  (All the tests pass, btw.)

As agreed, I didn't hunt down all potential bugs from unhandled reversed_type_impl, just the ones necessary to pass the tests.

In v2:
- fixed #7875
- added utility methods to abstract_type that handle reversed types under the hood
- dropped my tests (will rely on Nadav's instead)

nyh reviewed (approved):
Thanks, looks good. I'll test and if all goes well, commit.
We should remember that after this commit, still two potential problems remain:
1. Although you got rid of a couple of asserts() and replaced them by an exception (thanks!), I see there are still a few asserts() missing. I don't know how to trigger them now, but there's is potential that maliciously crafted junk in the CQL protocol can lead to them. Eventually we should get rid of these asserts, but it's not urgent now.
2. As you also also state, there may be additional existing code which checks is_set() and this patch doesn't fix and we didn't yet find in tests. Eventually we may want to look into that, but doesn't need to happen now.

psarna commented:
What's the status of this pull request? It hasn't been updated in a while and I see some merge conflicts.

dekimir commented:
@psarna It was merged and should have been closed automatically.

avikivity commented:
Merged in 176ff0238abba11cf0935cf09f71a363b25aab58. A bad rebase killed the merge commit, but the meat is there.
